input_network_encoder.cpp

```diff
@@ -0,0 +1,371 @@
+			break;
+		case DataBuffer::DATA_TYPE_REAL:
+			ERR_FAIL_COND_V_MSG(p_default_value.get_type() != Variant::FLOAT, UINT32_MAX, "The moveset initialization failed for" + p_name + " the specified data type is `REAL` but the default parameter is " + itos(p_default_value.get_type()));
+			break;
+		case DataBuffer::DATA_TYPE_POSITIVE_UNIT_REAL:
+			ERR_FAIL_COND_V_MSG(p_default_value.get_type() != Variant::FLOAT, UINT32_MAX, "The moveset initialization failed for" + p_name + " the specified data type is `REAL` but the default parameter is " + itos(p_default_value.get_type()));
+			break;
+		case DataBuffer::DATA_TYPE_UNIT_REAL:
+			ERR_FAIL_COND_V_MSG(p_default_value.get_type() != Variant::FLOAT, UINT32_MAX, "The moveset initialization failed for" + p_name + " the specified data type is `REAL` but the default parameter is " + itos(p_default_value.get_type()));
+			break;
+		case DataBuffer::DATA_TYPE_VECTOR2:
+			ERR_FAIL_COND_V_MSG(p_default_value.get_type() != Variant::VECTOR2, UINT32_MAX, "The moveset initialization failed for" + p_name + " the specified data type is `Vector2` but the default parameter is " + itos(p_default_value.get_type()));
+			break;
+		case DataBuffer::DATA_TYPE_NORMALIZED_VECTOR2:
+			ERR_FAIL_COND_V_MSG(p_default_value.get_type() != Variant::VECTOR2, UINT32_MAX, "The moveset initialization failed for" + p_name + " the specified data type is `Vector2` but the default parameter is " + itos(p_default_value.get_type()));
+			break;
+		case DataBuffer::DATA_TYPE_VECTOR3:
+			ERR_FAIL_COND_V_MSG(p_default_value.get_type() != Variant::VECTOR3, UINT32_MAX, "The moveset initialization failed for" + p_name + " the specified data type is `Vector3` but the default parameter is " + itos(p_default_value.get_type()));
+			break;
+		case DataBuffer::DATA_TYPE_NORMALIZED_VECTOR3:
+			ERR_FAIL_COND_V_MSG(p_default_value.get_type() != Variant::VECTOR3, UINT32_MAX, "The moveset initialization failed for" + p_name + " the specified data type is `Vector3` but the default parameter is " + itos(p_default_value.get_type()));
+			break;
+		case DataBuffer::DATA_TYPE_VARIANT:
+			/* No need to check variant, anything is accepted at this point.*/
+			break;
+	};
+
+	const uint32_t index = input_info.size();
+
+	input_info.resize(input_info.size() + 1);
+	input_info[index].name = p_name;
+	input_info[index].default_value = p_default_value;
+	input_info[index].data_type = p_type;
+	input_info[index].compression_level = p_compression_level;
+	input_info[index].comparison_floating_point_precision = p_comparison_floating_point_precision;
+
+	return index;
+}
+
+uint32_t InputNetworkEncoder::find_input_id(const StringName &p_name) const {
+	for (uint32_t i = 0; i < input_info.size(); i += 1) {
+		if (input_info[i].name == p_name) {
+			return i;
+		}
+	}
+	return INDEX_NONE;
```

Comment:
when I have such functions which are returning uncommon value to act as exception I always wonder how much do I expect for the person who is using this function to always check if that exception happened, couple thoughts:
- For sure such function documentation is always welcome
-  other thing is the usage of INDEX_NONE in situation when it's not exposed to GdScript... 
- For such ID thing I usually expect it will return -1 in case of error actually (since that value very clearly dont make any sense and it's very clearly intentional when debugging (some very big int feels like a pointer or smth)
- If I expect that its not super rare for user to check it I usually actually add new function called 'has_input_id' which is documented, and do an assertion inside find_input_id (like soft crash) with communicate that there is no such id and 'has_input_id' should be used - only by the fact  such function is on a list when one is reading codumention/help makes him aware that it can happen

Reply:
You are perfectly correct. Also I just realized `INDEX_NONE` is not exposed to GDS.